importers/rules.go

```diff
@@ -0,0 +1,256 @@
+
+	writer *blocks.MultiWriter
+}
+
+// RuleConfig is the config for the rule importer
+type RuleConfig struct {
+	Start        string
+	End          string
+	EvalInterval time.Duration
+	URL          string
+}
+
+// NewRuleImporter creates a new rule importer
+func NewRuleImporter(logger log.Logger, config RuleConfig) *RuleImporter {
+	return &RuleImporter{
+		config: config,
+	}
+}
+
+// Init initializes the rule importer which creates a new block writer
+// and creates an Prometheus API client
+func (importer *RuleImporter) Init() error {
+	// create new block writer
+	newBlockDir, err := ioutil.TempDir("", "rule_blocks")
+	if err != nil {
+		return err
+	}
+	importer.writer = blocks.NewMultiWriter(importer.logger, newBlockDir, importer.config.EvalInterval.Nanoseconds())
+
+	// create api client
+	config := api.Config{
+		Address: importer.config.URL,
+	}
+	c, err := api.NewClient(config)
+	if err != nil {
+		return err
+	}
+	importer.apiClient = v1.NewAPI(c)
+	return nil
+}
+
+// Close cleans up any open resources
+func (importer *RuleImporter) Close() error {
+	return importer.writer.Close()
+}
+
+// Parse parses the groups and rules from a list of rules files
+func (importer *RuleImporter) Parse(ctx context.Context, files []string) (errs []error) {
+	groups := make(map[string]*rules.Group)
+
+	for _, file := range files {
+		ruleGroups, errs := rulefmt.ParseFile(file)
+		if errs != nil {
+			return errs
+		}
+
+		for _, ruleGroup := range ruleGroups.Groups {
+			itv := importer.config.EvalInterval
+			if ruleGroup.Interval != 0 {
+				itv = time.Duration(ruleGroup.Interval)
+			}
+
+			rulez := make([]rules.Rule, 0, len(ruleGroup.Rules))
+			for _, r := range ruleGroup.Rules {
+				expr, err := parser.ParseExpr(r.Expr.Value)
+				if err != nil {
+					return []error{errors.Wrap(err, file)}
+				}
+
+				rulez = append(rulez, rules.NewRecordingRule(
+					r.Record.Value,
+					expr,
+					labels.FromMap(r.Labels),
+				))
+			}
+
+			groups[file+";"+ruleGroup.Name] = rules.NewGroup(rules.GroupOptions{
+				Name:     ruleGroup.Name,
+				File:     file,
+				Interval: itv,
+				Rules:    rulez,
+			})
+		}
+	}
+
+	importer.groups = groups
+	return errs
+}
+
+// ImportAll evaluates all the groups and rules and creates new time series
+// and stores in new blocks
+func (importer *RuleImporter) ImportAll(ctx context.Context) []error {
+	var errs = []error{}
+	for _, group := range importer.groups {
+		for _, rule := range group.Rules() {
+			err := importer.ImportRule(ctx, rule)
+			if err != nil {
+				errs = append(errs, err)
+			}
+		}
+	}
+	err := importer.CreateBlocks()
+	if err != nil {
+		errs = append(errs, err)
+	}
+	return errs
+}
+
+func (importer *RuleImporter) queryFn(ctx context.Context, q string, t time.Time) (promql.Vector, error) {
+	val, warnings, err := importer.apiClient.Query(ctx, q, t)
+	if err != nil {
+		return promql.Vector{}, err
+	}
+	if warnings != nil {
+		fmt.Fprint(os.Stderr, "warning api.Query:", warnings)
```

Comment:
Libraries shouldn't print.

Reply:
Do you have any recommendation on how I should handle the warnings?

Reply:
I logged the warnings instead.